:user-invalid triggering while typing date

pxlcoder opened the pull request:
#### 11990f0a5913ec3eb25802624dfb35975fea5a64
<pre>
:user-invalid triggering while typing date
<a href="https://bugs.webkit.org/show_bug.cgi?id=257999">https://bugs.webkit.org/show_bug.cgi?id=257999</a>
<a href="https://rdar.apple.com/110687369">rdar://110687369</a>

Reviewed by Tim Nguyen and Wenson Hsieh.

In order to dispatch input/change events for date/time inputs with multiple
editable fields, the element value is compared to the value derived from the
content of each field, every time the content of an individual field changes.

However, since the initial value of an element is a null string, and the
derived value is a sanitized string which defaults to the empty string,
comparing the two strings results in inequality. Consequently, an input event
is unexpectedly dispatched. Since the same machinery is involved in matching
`:user-invalid`, `:user-invalid` is also unexpectedly triggered as the user
types an incomplete date.

Fix by using `equalIgnoringNullity` to compare the two strings. This is correct
since `HTMLInputElement`&apos;s value uses `LegacyNullToEmptyString`.

Additionally, `:user-invalid` and `:user-valid` apply too early for date/time
inputs with editable fields. Specifically, the apply immediately once the value
of the input changes, not after the element loses focus following a value change.
This behavior also means that the &quot;change&quot; event fires too early. Since this
value is modified by typing, `:user-invalid` and `:user-valid` should only apply
after the input has lost focus, when matching for the first time. Fix by
reworking the event dispatch logic when the value is changed by typing.

Add additional test coverage for `input`, `change`, `:user-invalid`, and
`:user-valid` for date/time inputs.

* LayoutTests/fast/forms/date/date-editable-components/date-editable-components-input-and-change-events-expected.txt: Added.
* LayoutTests/fast/forms/date/date-editable-components/date-editable-components-input-and-change-events.html: Added.
* LayoutTests/fast/forms/date/date-editable-components/date-editable-components-keyboard-events-expected.txt:
* LayoutTests/fast/forms/date/date-editable-components/date-editable-components-keyboard-events.html:
* LayoutTests/fast/forms/date/date-editable-components/date-editable-components-user-invalid-expected.html: Added.
* LayoutTests/fast/forms/date/date-editable-components/date-editable-components-user-invalid-uncommitted-expected.html: Added.
* LayoutTests/fast/forms/date/date-editable-components/date-editable-components-user-invalid-uncommitted.html: Added.
* LayoutTests/fast/forms/date/date-editable-components/date-editable-components-user-invalid-with-empty-components-expected.html: Added.
* LayoutTests/fast/forms/date/date-editable-components/date-editable-components-user-invalid-with-empty-components.html: Added.
* LayoutTests/fast/forms/date/date-editable-components/date-editable-components-user-invalid.html: Added.
* LayoutTests/fast/forms/date/date-editable-components/date-editable-components-user-valid-expected.html: Added.
* LayoutTests/fast/forms/date/date-editable-components/date-editable-components-user-valid-uncommitted-expected.html: Added.
* LayoutTests/fast/forms/date/date-editable-components/date-editable-components-user-valid-uncommitted.html: Added.
* LayoutTests/fast/forms/date/date-editable-components/date-editable-components-user-valid.html: Added.
* LayoutTests/fast/forms/date/date-editable-components/date-multiple-fields-choose-default-value-after-set-value.html:
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/datetimelocal-editable-components-input-and-change-events-expected.txt: Added.
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/datetimelocal-editable-components-input-and-change-events.html: Added.
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/datetimelocal-editable-components-keyboard-events-expected.txt:
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/datetimelocal-editable-components-keyboard-events.html:
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/datetimelocal-editable-components-user-invalid-expected.html: Added.
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/datetimelocal-editable-components-user-invalid-uncommitted-expected.html: Added.
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/datetimelocal-editable-components-user-invalid-uncommitted.htm Added.
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/datetimelocal-editable-components-user-invalid-with-empty-components-expected.html: Added.
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/datetimelocal-editable-components-user-invalid-with-empty-components.html: Added.
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/datetimelocal-editable-components-user-invalid.html: Added.
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/datetimelocal-editable-components-user-valid-expected.html: Added.
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/datetimelocal-editable-components-user-valid-uncommitted-expected.html: Added.
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/datetimelocal-editable-components-user-valid-uncommitted.html: Added.
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/datetimelocal-editable-components-user-valid.html: Added.
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/datetimelocal-multiple-fields-choose-default-value-after-set-value.html:
* LayoutTests/fast/forms/month/month-editable-components/month-editable-components-input-and-change-events-expected.txt: Added.
* LayoutTests/fast/forms/month/month-editable-components/month-editable-components-input-and-change-events.html: Added.
* LayoutTests/fast/forms/month/month-editable-components/month-editable-components-keyboard-events-expected.txt:
* LayoutTests/fast/forms/month/month-editable-components/month-editable-components-keyboard-events.html:
* LayoutTests/fast/forms/month/month-editable-components/month-editable-components-user-invalid-expected.html: Added.
* LayoutTests/fast/forms/month/month-editable-components/month-editable-components-user-invalid-uncommitted-expected.html: Added.
* LayoutTests/fast/forms/month/month-editable-components/month-editable-components-user-invalid-uncommitted.html: Added.
* LayoutTests/fast/forms/month/month-editable-components/month-editable-components-user-invalid-with-empty-components-expected.html: Added.
* LayoutTests/fast/forms/month/month-editable-components/month-editable-components-user-invalid-with-empty-components.html: Added.
* LayoutTests/fast/forms/month/month-editable-components/month-editable-components-user-invalid.html: Added.
* LayoutTests/fast/forms/month/month-editable-components/month-editable-components-user-valid-expected.html: Added.
* LayoutTests/fast/forms/month/month-editable-components/month-editable-components-user-valid-uncommitted-expected.html: Added.
* LayoutTests/fast/forms/month/month-editable-components/month-editable-components-user-valid-uncommitted.html: Added.
* LayoutTests/fast/forms/month/month-editable-components/month-editable-components-user-valid.html: Added.
* LayoutTests/fast/forms/time/time-editable-components/time-editable-components-input-and-change-events-expected.txt: Added.
* LayoutTests/fast/forms/time/time-editable-components/time-editable-components-input-and-change-events.html: Added.
* LayoutTests/fast/forms/time/time-editable-components/time-editable-components-keyboard-events-expected.txt:
* LayoutTests/fast/forms/time/time-editable-components/time-editable-components-keyboard-events.html:
* LayoutTests/fast/forms/time/time-editable-components/time-editable-components-user-invalid-expected.html: Added.
* LayoutTests/fast/forms/time/time-editable-components/time-editable-components-user-invalid-uncommitted-expected.html: Added.
* LayoutTests/fast/forms/time/time-editable-components/time-editable-components-user-invalid-uncommitted.html: Added.
* LayoutTests/fast/forms/time/time-editable-components/time-editable-components-user-invalid-with-empty-components-expected.html: Added.
* LayoutTests/fast/forms/time/time-editable-components/time-editable-components-user-invalid-with-empty-components.html: Added.
* LayoutTests/fast/forms/time/time-editable-components/time-editable-components-user-invalid.html: Added.
* LayoutTests/fast/forms/time/time-editable-components/time-editable-components-user-valid-expected.html: Added.
* LayoutTests/fast/forms/time/time-editable-components/time-editable-components-user-valid-uncommitted-expected.html: Added.
* LayoutTests/fast/forms/time/time-editable-components/time-editable-components-user-valid-uncommitted.html: Added.
* LayoutTests/fast/forms/time/time-editable-components/time-editable-components-user-valid.html: Added.
* LayoutTests/fast/forms/time/time-editable-components/time-multiple-fields-choose-default-value-after-set-value.html:
* Source/WebCore/html/BaseDateAndTimeInputType.cpp:
(WebCore::BaseDateAndTimeInputType::didBlurFromControl):

When a date/time input with editable fields is focused, the actual focused
element will be one of it&apos;s editable fields. If one of the fields is blurred,
dispatch the &quot;change&quot; event if the value has changed since the previous &quot;change&quot;
event.

(WebCore::BaseDateAndTimeInputType::didChangeValueFromControl):

Dispatch only the input event when the input focused, following a change via
typing in one of the editable fields.

Canonical link: <a href="https://commits.webkit.org/272346@main">https://commits.webkit.org/272346@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7be488d5f1f8bf0c99e29a24ae90c983f455bde7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/31231 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/9908 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/32925 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/33737 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/28345 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/31998 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/12251 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/7165 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/27995 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/31568 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/8351 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/27908 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/7169 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/7346 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/27803 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/35078 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/28418 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/28260 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/33478 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/7394 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/5448 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/31315 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/9073 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/27597 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7366 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/8094 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/7920 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m reviewed (approved):
Looks good but we should also have basic test coverage that :user-valid/invalid does actually work for datetime inputs.

pxlcoder committed:
:user-invalid triggering while typing date
https://bugs.webkit.org/show_bug.cgi?id=257999
rdar://110687369

Reviewed by Tim Nguyen and Wenson Hsieh.

In order to dispatch input/change events for date/time inputs with multiple
editable fields, the element value is compared to the value derived from the
content of each field, every time the content of an individual field changes.

However, since the initial value of an element is a null string, and the
derived value is a sanitized string which defaults to the empty string,
comparing the two strings results in inequality. Consequently, an input event
is unexpectedly dispatched. Since the same machinery is involved in matching
`:user-invalid`, `:user-invalid` is also unexpectedly triggered as the user
types an incomplete date.

Fix by using `equalIgnoringNullity` to compare the two strings. This is correct
since `HTMLInputElement`'s value uses `LegacyNullToEmptyString`.

Additionally, `:user-invalid` and `:user-valid` apply too early for date/time
inputs with editable fields. Specifically, the apply immediately once the value
of the input changes, not after the element loses focus following a value change.
This behavior also means that the "change" event fires too early. Since this
value is modified by typing, `:user-invalid` and `:user-valid` should only apply
after the input has lost focus, when matching for the first time. Fix by
reworking the event dispatch logic when the value is changed by typing.

Add additional test coverage for `input`, `change`, `:user-invalid`, and
`:user-valid` for date/time inputs.

* LayoutTests/fast/forms/date/date-editable-components/date-editable-components-input-and-change-events-expected.txt: Added.
* LayoutTests/fast/forms/date/date-editable-components/date-editable-components-input-and-change-events.html: Added.
* LayoutTests/fast/forms/date/date-editable-components/date-editable-components-keyboard-events-expected.txt:
* LayoutTests/fast/forms/date/date-editable-components/date-editable-components-keyboard-events.html:
* LayoutTests/fast/forms/date/date-editable-components/date-editable-components-user-invalid-expected.html: Added.
* LayoutTests/fast/forms/date/date-editable-components/date-editable-components-user-invalid-uncommitted-expected.html: Added.
* LayoutTests/fast/forms/date/date-editable-components/date-editable-components-user-invalid-uncommitted.html: Added.
* LayoutTests/fast/forms/date/date-editable-components/date-editable-components-user-invalid-with-empty-components-expected.html: Added.
* LayoutTests/fast/forms/date/date-editable-components/date-editable-components-user-invalid-with-empty-components.html: Added.
* LayoutTests/fast/forms/date/date-editable-components/date-editable-components-user-invalid.html: Added.
* LayoutTests/fast/forms/date/date-editable-components/date-editable-components-user-valid-expected.html: Added.
* LayoutTests/fast/forms/date/date-editable-components/date-editable-components-user-valid-uncommitted-expected.html: Added.
* LayoutTests/fast/forms/date/date-editable-components/date-editable-components-user-valid-uncommitted.html: Added.
* LayoutTests/fast/forms/date/date-editable-components/date-editable-components-user-valid.html: Added.
* LayoutTests/fast/forms/date/date-editable-components/date-multiple-fields-choose-default-value-after-set-value.html:
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/datetimelocal-editable-components-input-and-change-events-expected.txt: Added.
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/datetimelocal-editable-components-input-and-change-events.html: Added.
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/datetimelocal-editable-components-keyboard-events-expected.txt:
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/datetimelocal-editable-components-keyboard-events.html:
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/datetimelocal-editable-components-user-invalid-expected.html: Added.
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/datetimelocal-editable-components-user-invalid-uncommitted-expected.html: Added.
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/datetimelocal-editable-components-user-invalid-uncommitted.htm Added.
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/datetimelocal-editable-components-user-invalid-with-empty-components-expected.html: Added.
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/datetimelocal-editable-components-user-invalid-with-empty-components.html: Added.
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/datetimelocal-editable-components-user-invalid.html: Added.
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/datetimelocal-editable-components-user-valid-expected.html: Added.
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/datetimelocal-editable-components-user-valid-uncommitted-expected.html: Added.
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/datetimelocal-editable-components-user-valid-uncommitted.html: Added.
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/datetimelocal-editable-components-user-valid.html: Added.
* LayoutTests/fast/forms/datetimelocal/datetimelocal-editable-components/datetimelocal-multiple-fields-choose-default-value-after-set-value.html:
* LayoutTests/fast/forms/month/month-editable-components/month-editable-components-input-and-change-events-expected.txt: Added.
* LayoutTests/fast/forms/month/month-editable-components/month-editable-components-input-and-change-events.html: Added.
* LayoutTests/fast/forms/month/month-editable-components/month-editable-components-keyboard-events-expected.txt:
* LayoutTests/fast/forms/month/month-editable-components/month-editable-components-keyboard-events.html:
* LayoutTests/fast/forms/month/month-editable-components/month-editable-components-user-invalid-expected.html: Added.
* LayoutTests/fast/forms/month/month-editable-components/month-editable-components-user-invalid-uncommitted-expected.html: Added.
* LayoutTests/fast/forms/month/month-editable-components/month-editable-components-user-invalid-uncommitted.html: Added.
* LayoutTests/fast/forms/month/month-editable-components/month-editable-components-user-invalid-with-empty-components-expected.html: Added.
* LayoutTests/fast/forms/month/month-editable-components/month-editable-components-user-invalid-with-empty-components.html: Added.
* LayoutTests/fast/forms/month/month-editable-components/month-editable-components-user-invalid.html: Added.
* LayoutTests/fast/forms/month/month-editable-components/month-editable-components-user-valid-expected.html: Added.
* LayoutTests/fast/forms/month/month-editable-components/month-editable-components-user-valid-uncommitted-expected.html: Added.
* LayoutTests/fast/forms/month/month-editable-components/month-editable-components-user-valid-uncommitted.html: Added.
* LayoutTests/fast/forms/month/month-editable-components/month-editable-components-user-valid.html: Added.
* LayoutTests/fast/forms/time/time-editable-components/time-editable-components-input-and-change-events-expected.txt: Added.
* LayoutTests/fast/forms/time/time-editable-components/time-editable-components-input-and-change-events.html: Added.
* LayoutTests/fast/forms/time/time-editable-components/time-editable-components-keyboard-events-expected.txt:
* LayoutTests/fast/forms/time/time-editable-components/time-editable-components-keyboard-events.html:
* LayoutTests/fast/forms/time/time-editable-components/time-editable-components-user-invalid-expected.html: Added.
* LayoutTests/fast/forms/time/time-editable-components/time-editable-components-user-invalid-uncommitted-expected.html: Added.
* LayoutTests/fast/forms/time/time-editable-components/time-editable-components-user-invalid-uncommitted.html: Added.
* LayoutTests/fast/forms/time/time-editable-components/time-editable-components-user-invalid-with-empty-components-expected.html: Added.
* LayoutTests/fast/forms/time/time-editable-components/time-editable-components-user-invalid-with-empty-components.html: Added.
* LayoutTests/fast/forms/time/time-editable-components/time-editable-components-user-invalid.html: Added.
* LayoutTests/fast/forms/time/time-editable-components/time-editable-components-user-valid-expected.html: Added.
* LayoutTests/fast/forms/time/time-editable-components/time-editable-components-user-valid-uncommitted-expected.html: Added.
* LayoutTests/fast/forms/time/time-editable-components/time-editable-components-user-valid-uncommitted.html: Added.
* LayoutTests/fast/forms/time/time-editable-components/time-editable-components-user-valid.html: Added.
* LayoutTests/fast/forms/time/time-editable-components/time-multiple-fields-choose-default-value-after-set-value.html:
* Source/WebCore/html/BaseDateAndTimeInputType.cpp:
(WebCore::BaseDateAndTimeInputType::didBlurFromControl):

When a date/time input with editable fields is focused, the actual focused
element will be one of it's editable fields. If one of the fields is blurred,
dispatch the "change" event if the value has changed since the previous "change"
event.

(WebCore::BaseDateAndTimeInputType::didChangeValueFromControl):

Dispatch only the input event when the input focused, following a change via
typing in one of the editable fields.

Canonical link: https://commits.webkit.org/272346@main
